File: backend/app/schemas/packaging.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import Field

from app.schemas.architect import LongText, ShortText
from app.schemas.base import Schema
from app.schemas.workspace import ProjectWorkspace

PackageBuildStatus = Literal["pending", "ready", "failed"]
ArchiveContent = Annotated[str, Field(min_length=1, max_length=8_000_000)]


class RepositoryStatistics(Schema):
    """Repository facts calculated from the completed project workspace."""

    total_files: Annotated[int, Field(ge=0)]
    total_folders: Annotated[int, Field(ge=1)]
    total_size: Annotated[int, Field(ge=0)]


class PackageManifestContext(Schema):
    """Read-only mission context used to describe a workspace export."""

    mission_summary: LongText
    organization_summary: LongText
    departments: Annotated[list[ShortText], Field(max_length=8)]
    generated_workers: Annotated[list[ShortText], Field(max_length=32)]
    total_tasks: Annotated[int, Field(ge=0, le=64)]
    completed_tasks: Annotated[int, Field(ge=0, le=64)]
    generated_artifacts: Annotated[int, Field(ge=0, le=256)]


class PackageManifest(Schema):
    """Human- and machine-readable summary embedded in every export bundle."""

    project_name: ShortText
    mission_summary: LongText
    organization_summary: LongText
    departments: Annotated[list[ShortText], Field(max_length=8)]
    generated_workers: Annotated[list[ShortText], Field(max_length=32)]
    total_tasks: Annotated[int, Field(ge=0, le=64)]
    completed_tasks: Annotated[int, Field(ge=0, le=64)]
    generated_artifacts: Annotated[int, Field(ge=0, le=256)]
    repository_statistics: RepositoryStatistics
    package_timestamp: datetime


class ProjectPackage(Schema):
    """Metadata for a ZIP package produced from one immutable workspace revision."""

    package_id: ShortText
    project_name: ShortText
    package_version: ShortText
    created_at: datetime
    total_files: Annotated[int, Field(ge=0)]
    total_size: Annotated[int, Field(ge=0)]
    build_status: PackageBuildStatus
    source_workspace_id: ShortText
    source_workspace_updated_at: datetime


class ExportBundle(Schema):
    """The archive payload together with its typed package metadata and manifest."""

    archive_base64: ArchiveContent
    archive_file_name: ShortText
    included_files: Annotated[list[ShortText], Field(max_length=300)]
    manifest: PackageManifest
    project_package: ProjectPackage


class ProjectPackagingRequest(Schema):
    """An existing workspace plus read-only context needed for its export manifest."""

    workspace: ProjectWorkspace
    manifest_context: PackageManifestContext

File: frontend/lib/api/packaging.ts
import type { ProjectWorkspace } from "@/lib/api/workspace";

export type PackageBuildStatus = "pending" | "ready" | "failed";

export type RepositoryStatistics = {
  total_files: number;
  total_folders: number;
  total_size: number;
};

/** Read-only organization context included in a generated package manifest. */
export type PackageManifestContext = {
  departments: string[];
  generated_artifacts: number;
  generated_workers: string[];
  mission_summary: string;
  organization_summary: string;
  total_tasks: number;
  completed_tasks: number;
};

export type PackageManifest = PackageManifestContext & {
  package_timestamp: string;
  project_name: string;
  repository_statistics: RepositoryStatistics;
};

/** Metadata retained for a package without storing the ZIP in browser session storage. */
export type ProjectPackage = {
  build_status: PackageBuildStatus;
  created_at: string;
  package_id: string;
  package_version: string;
  project_name: string;
  source_workspace_id: string;
  source_workspace_updated_at: string;
  total_files: number;
  total_size: number;
};

/** A transient ZIP payload with its durable package metadata and manifest. */
export type ExportBundle = {
  archive_base64: string;
  archive_file_name: string;
  included_files: string[];
  manifest: PackageManifest;
  project_package: ProjectPackage;
};

type ApiErrorPayload = { code?: string; detail?: string; message?: string };

export class ProjectPackagingApiError extends Error {
  readonly code: string;
  readonly status: number;

  constructor({
    code,
    message,
    status,
  }: Readonly<{ code: string; message: string; status: number }>) {
    super(message);
    this.name = "ProjectPackagingApiError";
    this.code = code;
    this.status = status;
  }
}

const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

/** Produces a ZIP from existing workspace files; it does not regenerate the workspace. */
export async function requestProjectPackage(
  input: Readonly<{
    manifestContext: PackageManifestContext;
    workspace: ProjectWorkspace;
  }>,
): Promise<ExportBundle> {
  let response: Response;

  try {
    response = await fetch(`${apiBaseUrl}/api/v1/packages`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        manifest_context: input.manifestContext,
        workspace: input.workspace,
      }),
    });
  } catch {
    throw new ProjectPackagingApiError({
      code: "packaging_unreachable",
      message: "Genesis could not reach the Project Packaging Engine. Please try again.",
      status: 0,
    });
  }

  if (!response.ok) {
    const payload = (await response.json().catch(() => ({}))) as ApiErrorPayload;
    throw new ProjectPackagingApiError({
      code: payload.code ?? "packaging_error",
      message: payload.message ?? payload.detail ?? "Genesis could not package the project.",
      status: response.status,
    });
  }

  return (await response.json()) as ExportBundle;
}
